src/tests/stencil_tests.cpp

```diff
@@ -0,0 +1,169 @@
+
+  // Disable color/zeta limit errors
+  auto crash_register = reinterpret_cast<uint32_t *>(PGRAPH_REGISTER_BASE + 0x880);
+  auto crash_register_pre_test = *crash_register;
+  *crash_register = crash_register_pre_test & (~0x800);
+  
+  // Enable color, disable depth/stencil, set depth/stencil test params
+  {
+    auto p = pb_begin();
+    p = pb_push1(p, NV097_SET_COLOR_MASK, 0x01010101);
+    p = pb_push1(p, NV097_SET_DEPTH_TEST_ENABLE, false);
+    p = pb_push1(p, NV097_SET_DEPTH_MASK, false);
+    p = pb_push1(p, NV097_SET_STENCIL_TEST_ENABLE, false);
+    p = pb_push1(p, NV097_SET_STENCIL_MASK, 0xFF);
+    p = pb_push1(p, NV097_SET_STENCIL_FUNC, 0x207);
+    p = pb_push1(p, NV097_SET_STENCIL_FUNC_REF, params.stencil_ref_value);
+    p = pb_push1(p, NV097_SET_STENCIL_FUNC_MASK, 0xFF);
+    p = pb_push1(p, NV097_SET_STENCIL_OP_FAIL, NV097_SET_STENCIL_OP_V_KEEP);
+    p = pb_push1(p, NV097_SET_STENCIL_OP_ZFAIL, NV097_SET_STENCIL_OP_V_KEEP);
+    p = pb_push1(p, NV097_SET_STENCIL_OP_ZPASS, params.stencil_op_zpass);
+    pb_end(p);
+  }
+
+  // Draw a red quad in the center of the screen
+  this->CreateGeometry(quadSize, 1, 0, 0);
+  host_.DrawArrays();
+
+  // Disable all masks except stencil, setup stencil test
+  {
+    auto p = pb_begin();
+    p = pb_push1(p, NV097_SET_COLOR_MASK, 0);
+    p = pb_push1(p, NV097_SET_STENCIL_TEST_ENABLE, params.stencil_test_enabled);
+    p = pb_push1(p, NV097_SET_DEPTH_MASK, false);
+    p = pb_push1(p, NV097_SET_DEPTH_TEST_ENABLE, params.depth_test_enabled);
+    p = pb_push1(p, NV097_SET_DEPTH_FUNC, NV097_SET_DEPTH_FUNC_V_ALWAYS);
+    pb_end(p);
+  }
+
+  // Draw a smaller quad to the zeta buffer
+  this->CreateGeometry(quadSize / 2.0f, 0, 0, 1);
+  host_.DrawArrays();
+
+  // Reenable masks, stencil test
+  {
+    auto p = pb_begin();
+    p = pb_push1(p, NV097_SET_COLOR_MASK, 0x01010101);
+    p = pb_push1(p, NV097_SET_DEPTH_TEST_ENABLE, false);
+    p = pb_push1(p, NV097_SET_DEPTH_MASK, false);
+    p = pb_push1(p, NV097_SET_STENCIL_TEST_ENABLE, true);
+    p = pb_push1(p, NV097_SET_STENCIL_FUNC, 0x202);
+    p = pb_push1(p, NV097_SET_STENCIL_OP_ZPASS, NV097_SET_STENCIL_OP_V_KEEP);
+    pb_end(p);
+  }
+
+  // Draw the first quad again, with a different color
+  this->CreateGeometry(quadSize, 0, 1, 0);
+  host_.DrawArrays();
+
+  pb_print("Stencil Op: %s\n", params.stencil_op_zpass_str);
+  pb_print("Stencil test: %s\n", params.stencil_test_enabled ? "Enabled" : "Disabled");
+  pb_print("Depth test: %s\n", params.depth_test_enabled ? "Enabled" : "Disabled");
+  pb_draw_text_screen();
+
+  std::string name = MakeTestName(params);
+  std::string z_name = name + "_ZB";
+  host_.FinishDraw(allow_saving_, output_dir_, name, z_name);
+
+  // Restore pgraph register 0x880
+  *crash_register = crash_register_pre_test;
+}
+
+void StencilTests::CreateGeometry(const float sideLength, const float r, const float g, const float b) {
+  auto fb_width = static_cast<float>(host_.GetFramebufferWidth());
```

Comment:
Nit/FYI: I added "F" variants of these when I got tired of casting, e.g., `GetFramebufferWidthF()`

Reply:
Fixed